Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+    ComponentData::ComponentData(AZ::EntityComponentIdPair pairId)
+        : m_pairId(pairId)
+    {
+        AZ::ComponentApplicationBus::BroadcastResult(
+            m_entity, &AZ::ComponentApplicationBus::Events::FindEntity, AZ::EntityId(pairId.GetEntityId()));
+        AZ::ComponentId componentId = pairId.GetComponentId();
+
+        m_component = m_entity->FindComponent(componentId);
+
+        m_componentName = AzToolsFramework::GetFriendlyComponentName(m_component);
+
+        AzToolsFramework::EditorRequestBus::BroadcastResult(
+            m_iconPath,
+            &AzToolsFramework::EditorRequestBus::Events::GetComponentEditorIcon,
+            m_component->GetUnderlyingComponentType(),
+            m_component);
+    }
+
+    ComponentModeSwitcher::ComponentModeSwitcher()
+    {
+        ViewportEditorModeNotificationsBus::Handler::BusConnect(GetEntityContextId());
+        EntityCompositionNotificationBus::Handler::BusConnect();
+        ToolsApplicationNotificationBus::Handler::BusConnect();
+
+
+        // Create the switcher
+        ViewportUi::ViewportUiRequestBus::EventResult(
+            m_switcherId,
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::CreateSwitcher,
+            ViewportUi::Alignment::TopLeft);
+
+        // Initial Transform button
+        ViewportUi::ViewportUiRequestBus::EventResult(
+            m_transformButtonId,
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::CreateSwitcherButton,
+            m_switcherId,
+            "../../../../Assets/Editor/Icons/Components/Transform.svg",
+            "Transform");
+
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::SetSwitcherActiveButton,
+            m_switcherId,
+            m_transformButtonId);
+
+        // Handler
+        auto handlerFunc = [this](ViewportUi::ButtonId buttonId)
+        {
+            ViewportUi::ViewportUiRequestBus::Event(
+                ViewportUi::DefaultViewportId,
+                &ViewportUi::ViewportUiRequestBus::Events::SetSwitcherActiveButton,
+                m_switcherId,
+                buttonId);
+
+            ActivateComponentMode(buttonId);
+        };
+
+        m_handler = AZ::Event<ViewportUi::ButtonId>::Handler(handlerFunc);
+        // Registers the SwitcherEventHandler which is defined in ComponentModeSwitcher
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::RegisterSwitcherEventHandler,
+            m_switcherId,
+            m_handler);
+    }
+
+    ComponentModeSwitcher::~ComponentModeSwitcher()
+    {
+        ViewportEditorModeNotificationsBus::Handler::BusDisconnect();
+        EntityCompositionNotificationBus::Handler::BusDisconnect();
+        ToolsApplicationNotificationBus::Handler::BusDisconnect();
```

Comment:
update: It's good practice to disconnect in the reverse order in which you connected, so this would be...

```
ToolsApplicationNotificationBus::Handler::BusDisconnect();
EntityCompositionNotificationBus::Handler::BusDisconnect();
ViewportEditorModeNotificationsBus::Handler::BusDisconnect();
```